docs/K8sGateKeeper.mdx

```diff
@@ -0,0 +1,377 @@
+```
+
+Assume `accessWithWildcard(r.obj.Request.Object.Object.Spec.Template.Spec.Containers , "*", "Image")`returns `["a:b", "c:d", "e:f", "g:h"]` then because splits supports variable-length argument, and splits operation is applied on each element, and eventualy element whose index is 1 will be selected and return, so `split(accessWithWildcard(r.obj.Request.Object.Object.Spec.Template.Spec.Containers , "*", "Image"),":",1)` returns `["b","d","f","h"]`. And `contain(split(accessWithWildcard(r.obj.Request.Object.Object.Spec.Template.Spec.Containers , "*", "Image"),":",1) , p.obj)` returns whether`p.obj` is contained in `["b","d","f","h"]`
+
+#### 4.2.1.2 Type conversion functions
+- ParseFloat(): parse an integer to a float. (It is because that any number in comparsion must be converted into float).
+- ToString(): convert an object to string. This object must have a basic type of string. (for example, an object of type `XXX` when there is a statement `type XXX string`)
+- IsNil(): return whether the parameter is nil
+
+## 5. Advanced Settings
+
+### 5.1 About Certificates
+
+In k8s, it is mandatory that a webhook should use HTTPS. There are two approaches to achieve that:
+- Use self-signed certificates(examples in this repo use this method )
+- Use a normal certificate
+
+#### 5.1.1 Self-signed certificates
+Using a self-signed certificate means that the CA issuing the certificate is not one of the well-known CAs, therefore you must let k8s know this CA. 
+
+Current the example in this repo uses a self-made CA, whose private key and certificate is stored in `config/certificate/ca.crt` and `config/certificate/ca.key`. Certificate for the webhook is `config/certificate/server.crt`, issued by the self-made CA. The domains of this certificate is "webhook.domain.local"(for external webhook) and "casbin-webhook-svc.default.svc"(for internal webhook)
+
+Information about CA is passed to k8s via webhook configuration files. Both `config/webhook_external.yaml` and `config/webhook_internal.yaml` have a field called "CABundle", whose content is base64 encoded string of the  certificate of the CA.
+
+In case that you need to change the certificate/domain (for example, maybe you want to put this webhook into another namespace of k8s while using internal webhook; or maybe you want to change a domain while using external webhook), the following procedures should be taken:
+
+1. Generate a new CA
+ 
+Generate the private key for the fake CA
+```
+openssl genrsa -des3 -out ca.key 2048
+```
+
+Remove the password protection of the private key.
+``` 
+openssl rsa -in ca.key -out ca.key
+```
+
+2. Generate a private key for webhook server
+
+```
+openssl genrsa -des3 -out server.key 2048
+openssl rsa -in server.key  -out server.key 
+```
+3. Use the self-generate CA to sign the certificate for webhook
+
+
+Copy your system's openssl config file for temporary use. You can use `openssl version -a` to find out the location of the config file.
```

Comment:
It's better to use the filename of the config file instead of just using `the config file`.